### PR TITLE
Annotate NullableWalker

### PR DESCRIPTION
**Pull request by cston**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -51,11 +54,6 @@ internal sealed class VariableState
         /// </summary>
         private readonly PooledDictionary<Symbol, TypeSymbolWithAnnotations> _variableTypes = PooledDictionary<Symbol, TypeSymbolWithAnnotations>.GetInstance();
 
-        /// <summary>
-        /// The current source assembly.
-        /// </summary>
-        private readonly SourceAssemblySymbol _sourceAssembly;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >_sourceAssembly [](start = 46, length = 15)

Probably good to include this change with next opportune NullableWalker change into `master`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -5,6 +5,9 @@
 #define REFERENCE_STATE
 #endif
 
+#nullable enable
+#pragma warning disable 8600
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: How did things look with W warnings?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -3172,9 +3172,9 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
                     break;
 
                 case ConversionKind.AnonymousFunction:
-                    if (operandOpt.Kind == BoundKind.Lambda)
+                    if (operandOpt!.Kind == BoundKind.Lambda)
                     {
-                        var lambda = (BoundLambda)operandOpt;
+                        var lambda = (BoundLambda)operandOpt!;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >operandOpt! [](start = 50, length = 11)

This makes me wonder if we should make `!` sticky. I don't recall if that's still on the open issues list. What do you think?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, added to https://github.com/dotnet/roslyn/pull/31703.

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, added to [https://github.com/dotnet/roslyn/pull/31703](https://github.com/dotnet/roslyn/pull/31703).

---
In reply to: [241227466](https://github.com/dotnet/roslyn/pull/31719#discussion_r241227466) [](ancestors = 241227466)

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I agree. Logged [https://github.com/dotnet/roslyn/issues/31732](https://github.com/dotnet/roslyn/issues/31732).

---
In reply to: [241228773](https://github.com/dotnet/roslyn/pull/31719#discussion_r241228773) [](ancestors = 241228773)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: There were 43 W warnings in the file.

---
In reply to: [241228286](https://github.com/dotnet/roslyn/pull/31719#discussion_r241228286) [](ancestors = 241228286)

File: build/Targets/Packages.props [hunk 4]
```diff
@@ -76,10 +76,10 @@
     <MicrosoftMSXMLVersion>8.0.0.0-alpha</MicrosoftMSXMLVersion>
     <!-- Using a private build of Microsoft.Net.Test.SDK to work around issue https://github.com/Microsoft/vstest/issues/1764 -->
     <MicrosoftNETTestSdkVersion>15.9.0-dev2</MicrosoftNETTestSdkVersion>
-    <MicrosoftNetCompilersVersion>2.11.0-beta1-63431-04</MicrosoftNetCompilersVersion>
+    <MicrosoftNetCompilersVersion>3.0.0-beta2-18612-03</MicrosoftNetCompilersVersion>
```
Comment (jaredpar) [inline build/Targets/Packages.props]: Where did you get that version from?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 5]
```diff
@@ -860,7 +859,7 @@ private void EnterParameter(ParameterSymbol parameter, TypeSymbolWithAnnotations
 
 #region Visitors
 
-        public override BoundNode VisitIsPatternExpression(BoundIsPatternExpression node)
+        public override BoundNode? VisitIsPatternExpression(BoundIsPatternExpression node)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >BoundNode? [](start = 24, length = 10)

It looks like we always return parameter that isn't nullable. Why was this change necessary?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 6]
```diff
@@ -2875,7 +2874,7 @@ private TypeSymbolWithAnnotations GetAdjustedResult(TypeSymbolWithAnnotations ty
             return type;
         }
 
-        private static Symbol AsMemberOfResultType(TypeSymbolWithAnnotations resultType, Symbol symbol)
+        private static Symbol? AsMemberOfResultType(TypeSymbolWithAnnotations resultType, Symbol symbol)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Symbol? [](start = 23, length = 7)

It feels like we shouldn't add a ```?``` here, given that the ```symbol``` is not nullable.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 7]
```diff
@@ -2885,7 +2884,7 @@ private static Symbol AsMemberOfResultType(TypeSymbolWithAnnotations resultType,
             return AsMemberOfType(containingType, symbol);
         }
 
-        private static Symbol AsMemberOfType(NamedTypeSymbol containingType, Symbol symbol)
+        private static Symbol? AsMemberOfType(NamedTypeSymbol containingType, Symbol symbol)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Symbol symbol [](start = 78, length = 13)

```Symbol?```?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 8]
```diff
@@ -2016,14 +2015,14 @@ public override BoundNode VisitCall(BoundCall node)
                 if (!receiverType.IsNull)
                 {
                     // Update method based on inferred receiver type.
-                    method = (MethodSymbol)AsMemberOfResultType(receiverType, method);
+                    method = (MethodSymbol)AsMemberOfResultType(receiverType, method)!;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >(MethodSymbol)AsMemberOfResultType(receiverType, method)! [](start = 29, length = 57)

Should probably make return of AsMemberOfResultType not nullable instead.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 9]
```diff
@@ -3378,15 +3379,16 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
...
         {
+            Debug.Assert((object)target != null || assignmentKind != AssignmentKind.Argument);
             HashSet<DiagnosticInfo> useSiteDiagnostics = null;
             var conversion = _conversions.ClassifyStandardConversion(null, operandType.TypeSymbol, targetType.TypeSymbol, ref useSiteDiagnostics);
             if (!conversion.Exists)
             {
                 if (assignmentKind == AssignmentKind.Argument)
                 {
-                    ReportNullabilityMismatchInArgument(node, operandType.TypeSymbol, target, targetType.TypeSymbol);
+                    ReportNullabilityMismatchInArgument(node, operandType.TypeSymbol, target!, targetType.TypeSymbol);
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >`target!`

Considering the assert above, this shouldn't be required?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 10]
```diff
@@ -2016,14 +2015,14 @@ public override BoundNode VisitCall(BoundCall node)
...
-                    method = (MethodSymbol)AsMemberOfResultType(receiverType, method);
+                    method = (MethodSymbol)AsMemberOfResultType(receiverType, method)!;
                 }
 
                 ImmutableArray<RefKind> refKindsOpt = node.ArgumentRefKindsOpt;
                 (ImmutableArray<BoundExpression> arguments, ImmutableArray<Conversion> conversions) = RemoveArgumentConversions(node.Arguments, refKindsOpt);
                 ImmutableArray<int> argsToParamsOpt = node.ArgsToParamsOpt;
 
-                method = VisitArguments(node, arguments, refKindsOpt, method.Parameters, argsToParamsOpt, node.Expanded, node.InvokedAsExtensionMethod, conversions, method);
+                method = VisitArguments(node, arguments, refKindsOpt, method.Parameters, argsToParamsOpt, node.Expanded, node.InvokedAsExtensionMethod, conversions, method)!;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >method = VisitArguments(node, arguments, refKindsOpt, method.Parameters, argsToParamsOpt, node.Expanded, node.InvokedAsExtensionMethod, conversions, method)!; [](start = 16, length = 158)

It feels like we can benefit from an attribute saying the result is not nullable if some parameter is not nullable.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 11]
```diff
@@ -3213,7 +3213,8 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
                             checkConversion: false,
                             fromExplicitCast: false,
                             useLegacyWarnings,
-                            assignmentKind);
+                            assignmentKind,
+                            target);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >target [](start = 28, length = 6)

It doesn't look like this change is backed by any unit-test.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 12]
```diff
@@ -3230,7 +3231,7 @@ private bool HasTopLevelNullabilityConversion(TypeSymbolWithAnnotations source,
 
                         // method return type -> conversion "to" type
                         // May be distinct from method return type for Nullable<T>.
-                        operandType = ClassifyAndApplyConversion(operandOpt ?? node, TypeSymbolWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType), operandType, useLegacyWarnings, assignmentKind, target: null);
+                        operandType = ClassifyAndApplyConversion(operandOpt ?? node, TypeSymbolWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType), operandType, useLegacyWarnings, assignmentKind, target);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >operandType = ClassifyAndApplyConversion(operandOpt ?? node, TypeSymbolWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType), operandType, useLegacyWarnings, assignmentKind, target); [](start = 24, length = 204)

It doesn't look like this change is backed by a unit-test.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 10 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, once such an attribute is supported.

---
In reply to: [241506605](https://github.com/dotnet/roslyn/pull/31719#discussion_r241506605) [](ancestors = 241506605)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 13]
```diff
@@ -79,20 +77,20 @@ internal sealed class VariableState
...
         /// signature when _member is a lambda and type is inferred from MethodTypeInferrer.
         /// </summary>
-        private readonly MethodSymbol _methodSignatureOpt;
+        private readonly MethodSymbol? _methodSignatureOpt;
 
         /// <summary>
         /// Types from return expressions. Used when inferring lambda return type in MethodTypeInferrer.
         /// </summary>
-        private readonly ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)> _returnTypes;
+        private readonly ArrayBuilder<(RefKind, TypeSymbolWithAnnotations)>? _returnTypes;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Slightly surprising, should this be named `_returnTypesOpt`?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 14]
```diff
@@ -105,7 +103,7 @@ internal sealed class VariableState
         /// <summary>
         /// Instances being constructed.
         /// </summary>
-        private PooledDictionary<BoundExpression, ObjectCreationPlaceholderLocal> _placeholderLocals;
+        private PooledDictionary<BoundExpression, ObjectCreationPlaceholderLocal>? _placeholderLocals;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >a [](start = 98, length = 1)

Same comment.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 15]
```diff
@@ -529,7 +526,7 @@ private bool ReportNullableAssignmentIfNecessary(BoundExpression value, TypeSymb
             {
                 ReportDiagnostic(ErrorCode.WRN_NullReferenceArgument, value.Syntax,
                     new FormattedSymbol(target, SymbolDisplayFormat.ShortFormat),
-                    new FormattedSymbol(target.ContainingSymbol, SymbolDisplayFormat.MinimallyQualifiedFormat));
+                    new FormattedSymbol(target!.ContainingSymbol, SymbolDisplayFormat.MinimallyQualifiedFormat));
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Rather than a warning suppression, would it be better to `Debug.Assert` before?

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, although the asserts will fail currently. See [https://github.com/dotnet/roslyn/pull/31771](https://github.com/dotnet/roslyn/pull/31771) for a fix and the assert.

---
In reply to: [241893107](https://github.com/dotnet/roslyn/pull/31719#discussion_r241893107) [](ancestors = 241893107)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 13 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Renamed in [https://github.com/dotnet/roslyn/pull/31771](https://github.com/dotnet/roslyn/pull/31771) rather than here since it will be a while before this branch can be merged into master.

---
In reply to: [241890092](https://github.com/dotnet/roslyn/pull/31719#discussion_r241890092) [](ancestors = 241890092)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 16]
```diff
@@ -954,7 +951,7 @@ private void VisitPattern(BoundExpression expression, TypeSymbolWithAnnotations
             }
         }
 
-        protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement node)
+        protected override BoundNode? VisitReturnStatementNoAdjust(BoundReturnStatement node)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: More general question for these: once nullable expands further in the code-base, these overrides are going to need to match in return nullability, right?

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Specifically, even for the places that we currently have returning `BoundNode` in this file, they'll have to match the parent.

---
In reply to: [241927757](https://github.com/dotnet/roslyn/pull/31719#discussion_r241927757) [](ancestors = 241927757)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 17]
```diff
@@ -4766,15 +4759,15 @@ internal void EnsureCapacity(int capacity)
                 {
                     if (slot < Capacity)
                     {
-                        return _state[slot];
+                        return _state![slot];
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >_state! [](start = 31, length = 7)

Consider encapsulating `_state` through a private property to hide the fact that it could be null if not initialized.

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Wrapping this in a property seems like a bit of overhead for such a small type. And 3 of the 5 methods that use the field have to handle a null value.

---
In reply to: [241929195](https://github.com/dotnet/roslyn/pull/31719#discussion_r241929195) [](ancestors = 241929195)

**Comment by cston** [comment]
>                             target);

This change is from master.

---
In reply to: [448321059](https://github.com/dotnet/roslyn/pull/31719#issuecomment-448321059) [](ancestors = 448321059)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3416 in 732b5e5. [](commit_id = 732b5e5139eebfc1f15491b3f873433c67d05cad, deletion_comment = False)

**Comment by cston** [comment]
>                         operandType = ClassifyAndApplyConversion(operandOpt ?? node, TypeSymbolWithAnnotations.Create(conversion.BestUserDefinedConversionAnalysis.ToType), operandType, useLegacyWarnings, assignmentKind, target);

From master.

---
In reply to: [448321324](https://github.com/dotnet/roslyn/pull/31719#issuecomment-448321324) [](ancestors = 448321324)

---
Refers to: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:3454 in 732b5e5. [](commit_id = 732b5e5139eebfc1f15491b3f873433c67d05cad, deletion_comment = False)
